python/graphstorm/model/lm_embed.py

```diff
@@ -28,109 +29,200 @@
-        There is a one-one correspondence between the language models stored in
-        lm_models and the (lm_ntypes, lm_node_feats) tuples in lm_models_info.
-        The lm_ntypes stores the node types that share the same language model
-        and the lm_node_feats stores the corresponding language model computation
-        related node features (i.e., token ids, attention masks, etc.)
-    use_cache: bool
-        If true, use the embeddings in lm_emb_cache
+    g : DistGraph
+        The distributed graph.
+    node_lm_configs:
+        A list of language model configurations.
+    num_train: int
+        Number of trainable texts
+    lm_infer_batch_size: int
+        Batch size used for computing text embeddings for static lm model
     """
-    lm_feats = {}
-
-    # Get the device from lm_models
-    # The cached BERT embedding should be moved to the same device
-    # as lm_models.
-    dev = next(lm_models[0].parameters()).device
-    if use_cache:
-        # No bert training, Get cached LM embedding
-        # Note: self.lm_emb_cache is initialized by calling warmup
-        for ntype, idx in input_nodes.items():
-            if ntype in lm_emb_cache:
-                lm_feats[ntype] = lm_emb_cache[ntype][idx].to(dev)
-    else:
-        # TODO: Release the bert cache properly
-        #       This may need support from DistDGL
-        # Need bert training
-        for (lm_ntypes, lm_node_feats), lm_model \
-            in zip(lm_models_info, lm_models):
-            input_ntypes = []
-            input_lm_feats = {}
+    def __init__(self, g, node_lm_configs, num_train, lm_infer_batch_size):
+        super(LMModels, self).__init__()
+        # More than one node type may share the same BERT model.
+        # To avoid duplicate BERT models, we save add the BERT model
+        # in the module dict once, whose key is the node types of the BERT model.
+        # We then maintain a mapping from the node type to the key
+        # to help find the right BERT model for a node type.
+        self._lm_map = {}
+        self._lm_models = nn.ModuleDict()
+        self._lm_node_feats = {}
+        for lm_config in node_lm_configs:
+            lm_model = init_lm_model(lm_config,
+                                     num_train=num_train,
+                                     lm_infer_batch_size=lm_infer_batch_size)
+            # A list of node types sharing the same lm model
+            lm_ntypes = lm_config["node_types"]
+            lm_node_feats = get_lm_node_feats(g, lm_model, lm_ntypes)
+            for ntype, feats in lm_node_feats.items():
+                assert ntype not in self._lm_node_feats, \
+                        f"More than one BERT model runs on Node {ntype}."
+                self._lm_node_feats[ntype] = feats
+            key = str(lm_ntypes)
+            self._lm_models[key] = lm_model
             for ntype in lm_ntypes:
+                self._lm_map[ntype] = key
```

Comment:
What is this? A map from ntype to ntype?

Reply:
a mapping from ntype to the key in `_lm_models`. The key of `_lm_models` is not ntype. a LM model can be used by multiple node types. The key is the string that represents all node types using the LM model.

Reply:
Shall we sort the ntypes before creating the key?